File: gdscript/src/main/java/gdscript/GdIcon.java
package gdscript;

import com.intellij.openapi.util.IconLoader;
import com.intellij.ui.RowIcon;
import com.intellij.util.IconUtil;

import javax.swing.*;
import java.util.HashMap;

public class GdIcon {

    public static final Icon FILE = IconLoader.getIcon("icons/file.png", GdIcon.class);

    // Godot editor icons
    @Deprecated
    public static final Icon OBJECT = IconLoader.getIcon("icons/godot_editor/Object.svg", GdIcon.class);
    public static final String METHOD_MARKER = "Forward-sq";
    public static final String VAR_MARKER = "KeyBezierSelected";
    public static final String CONST_MARKER = "KeyXform";

    public static HashMap<String, Icon> editorIcons = new HashMap<>();

    public static Icon getEditorIcon(String className) {
        Icon icon = editorIcons.get(className);
        if (icon == null) {
            try {
                Icon loaded = IconLoader.getIcon(String.format("icons/godot_editor/%s.svg", className), GdIcon.class);
                if (loaded.getIconHeight() > 1) {
                    loaded = IconUtil.toSize(loaded, 16, 16);
                    editorIcons.put(className, loaded);
                } else {
                    editorIcons.put(className, OBJECT);
                }
            } catch (Exception e) {
                editorIcons.put(className, OBJECT);
            }
        }

        return editorIcons.get(className);
    }

}

File: gdscript/src/main/kotlin/gdscript/GdIcon.kt
package gdscript

import GdScriptPluginIcons
import com.intellij.openapi.diagnostic.thisLogger
import com.intellij.openapi.util.IconLoader
import org.jetbrains.annotations.ApiStatus
import javax.swing.Icon

object GdIcon {
    private var editorIcons = HashMap<String, Icon>()

    @ApiStatus.Obsolete // todo: find a way to get rid of this completely and also all unused Godot editor icons
    fun getEditorIcon(className: String): Icon {
        val icon = editorIcons[className]
        if (icon == null) {
            try {
                val loaded = IconLoader.getIcon(
                    String.format("icons/godot_editor/%s.svg", className),
                    GdIcon::class.java
                )
                if (loaded.iconHeight > 1) {
                    //loaded = IconUtil.toSize(loaded, 16, 16)
                    // todo: determine not needed icons and remove them
                    // rework icons to be required size without scaling
                    editorIcons[className] = loaded
                } else {
                    editorIcons[className] = GdScriptPluginIcons.Icons.BackupIcon
                }
            } catch (e: Exception) {
                thisLogger().error("Unable to load editor icon for $className. Using default one.", e)
                editorIcons[className] = GdScriptPluginIcons.Icons.BackupIcon
            }
        }

        return editorIcons[className]!!
    }
}
